File: 79_Word Search.cpp
//dfs
bool haxi[200][200];
class Solution {
public:
    int move[4][2] = {{1,0},{0,1},{-1,0},{0,-1}};
    int m,n;
    bool found;
    string word;
    void dfs(int i,int j,int wordpos,vector<vector<char>>& board)
    {
        if(found) return;
        if(wordpos==word.size())
        {
            found = true;
            return;
        }
        int nexti,nextj;
        for(int d = 0;d<4;d++)
        {
            nexti = i+move[d][0];
            nextj = j+move[d][1];
            if(nexti>=0 &&nexti<m &&nextj>=0&&nextj<n &&board[nexti][nextj]==word[wordpos] && 
               !haxi[nexti][nextj])
            {
                haxi[nexti][nextj] = 1;
                dfs(nexti,nextj,wordpos+1,board);
                haxi[nexti][nextj] = 0;
                if(found) return;
            }
        } 

    }
    bool exist(vector<vector<char>>& board, string word) {
        this->word = word;
        memset(haxi,0,sizeof(haxi));
        found = false;
        m = board.size();
        n = board[0].size();
        for(int i = 0;i<m;i++)
        {
            for(int j =0;j<n;j++)
            {
                if(board[i][j]==word[0])
                {
                    haxi[i][j] = 1;
                    dfs(i,j,1,board);
                    haxi[i][j] = 0;
                    if(found) return true;
                }
            }
        }
        return false;
    }
};

File: 79_Word Search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.board = board
        self.word = word
        self.m = len(board)
        self.n = len(board[0])
        self.path = [[0 for j in range(self.n)] for i in range(self.m)]
        self.directions = [[0,1],[1,0],[0,-1],[-1,0]]
        pos = 0
        while pos<self.m*self.n:
            i = pos//self.n
            j = pos%self.n
            pos+=1
            if board[i][j]!=word[0]:
                continue
            if self.gao(i,j,0):
                return True
        return False
    def gao(self,i,j,count):
        #print("%d %d %d" %(i,j,count))
        if i<0 or j<0 or i>=self.m or j>=self.n or self.path[i][j]==1 or self.board[i][j]!=self.word[count]:
            return False
        self.path[i][j] = 1
        if count==len(self.word)-1:
            return True
        for d in self.directions:
            if self.gao(i+d[0],j+d[1],count+1):
                return True
        self.path[i][j] = 0
        return False
            
        
